File: benches/shapes/shapes_baseline.cpp
// CPU baselines for the fir/conv2d Mapal shapes, size-parameterized (S29 scale-up).
// Build: clang++ -O3 -march=native -ffp-contract=fast -std=c++17 shapes_baseline.cpp -o shapes_baseline -pthread
#include <algorithm>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <functional>
#include <thread>
#include <vector>

static unsigned thread_width() {
    if (const char* value = std::getenv("THREADS")) {
        int threads = std::atoi(value);
        if (threads >= 1) return static_cast<unsigned>(threads);
    }

    long quota = -1;
    if (FILE* file = std::fopen("/sys/fs/cgroup/cpu.max", "r")) {
        char value[64] = {};
        long period = 0;
        if (std::fscanf(file, "%63s %ld", value, &period) == 2 &&
            std::strcmp(value, "max") != 0 && period > 0)
            quota = (std::atol(value) + period - 1) / period;
        std::fclose(file);
    }
    if (quota < 0) {
        long value = -1, period = -1;
        if (FILE* file = std::fopen("/sys/fs/cgroup/cpu/cpu.cfs_quota_us", "r")) {
            if (std::fscanf(file, "%ld", &value) != 1) value = -1;
            std::fclose(file);
        }
        if (FILE* file = std::fopen("/sys/fs/cgroup/cpu/cpu.cfs_period_us", "r")) {
            if (std::fscanf(file, "%ld", &period) != 1) period = -1;
            std::fclose(file);
        }
        if (value > 0 && period > 0) quota = (value + period - 1) / period;
    }

    unsigned threads = std::thread::hardware_concurrency();
    if (threads == 0) threads = 1;
    if (quota > 0) threads = std::min(threads, static_cast<unsigned>(quota));
    return threads;
}

static void fir_range(const std::vector<float>& x, const std::vector<float>& w,
                      std::vector<float>& y, size_t begin, size_t end) {
    for (size_t t = begin; t < end; ++t) {
        float acc = 0.0f;
        for (size_t k = 0; k < w.size(); ++k) acc += w[k] * x[t + k];
        y[t] = acc;
    }
}

static void conv_range(const std::vector<float>& img, const std::vector<float>& w,
                       std::vector<float>& out, size_t side, size_t row_begin, size_t row_end) {
    const size_t stride = side + 2;
    for (size_t i = row_begin; i < row_end; ++i) {
        for (size_t j = 0; j < side; ++j) {
            float acc = 0.0f;
            for (size_t k = 0; k < 9; ++k)
                acc += w[k] * img[(i + k / 3) * stride + j + k % 3];
            out[i * side + j] = acc;
        }
    }
}

template <typename Work>
static void run_iters(int iters, Work work) {
    for (int iter = 0; iter < iters; ++iter) {
        const auto start = std::chrono::steady_clock::now();
        work();
        const double ms =
            std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - start)
                .count();
        std::printf("iter ms=%.6f\n", ms);
    }
}

static void run_fir(bool multithreaded, int iters, size_t n) {
    std::vector<float> x(n + 63), w(64), y(n);
    for (size_t t = 0; t < x.size(); ++t)
        x[t] = static_cast<float>(static_cast<int>((t * 7 + 13) % 101) - 50);
    for (size_t k = 0; k < w.size(); ++k)
        w[k] = static_cast<float>(static_cast<int>((k * 5 + 3) % 31) - 15);

    const unsigned threads = std::min<unsigned>(thread_width(), y.size());
    run_iters(iters, [&] {
        if (!multithreaded) {
            fir_range(x, w, y, 0, y.size());
            return;
        }
        std::vector<std::thread> workers;
        workers.reserve(threads);
        for (unsigned lane = 0; lane < threads; ++lane) {
            const size_t begin = lane * y.size() / threads;
            const size_t end = (lane + 1) * y.size() / threads;
            workers.emplace_back(fir_range, std::cref(x), std::cref(w), std::ref(y), begin, end);
        }
        for (auto& worker : workers) worker.join();
    });
    std::printf("%.9g\n%.9g\n", static_cast<double>(y.front()), static_cast<double>(y.back()));
}

static void run_conv(bool multithreaded, int iters, size_t side) {
    const size_t stride = side + 2;
    std::vector<float> img(stride * stride), w(9), out(side * side);
    for (size_t t = 0; t < img.size(); ++t)
        img[t] = static_cast<float>(static_cast<int>((t * 7 + 13) % 101) - 50);
    for (size_t k = 0; k < w.size(); ++k)
        w[k] = static_cast<float>(static_cast<int>((k * 5 + 3) % 31) - 15);

    const unsigned threads = std::min<unsigned>(thread_width(), side);
    run_iters(iters, [&] {
        if (!multithreaded) {
            conv_range(img, w, out, side, 0, side);
            return;
        }
        std::vector<std::thread> workers;
        workers.reserve(threads);
        for (unsigned lane = 0; lane < threads; ++lane) {
            const size_t begin = lane * side / threads;
            const size_t end = (lane + 1) * side / threads;
            workers.emplace_back(conv_range, std::cref(img), std::cref(w), std::ref(out), side, begin, end);
        }
        for (auto& worker : workers) worker.join();
    });
    std::printf("%.9g\n%.9g\n", static_cast<double>(out.front()), static_cast<double>(out.back()));
}

int main(int argc, char** argv) {
    if (argc < 4 || argc > 5 || (std::strcmp(argv[1], "fir") != 0 && std::strcmp(argv[1], "conv2d") != 0) ||
        (std::strcmp(argv[2], "1t") != 0 && std::strcmp(argv[2], "mt") != 0)) {
        std::fprintf(stderr, "usage: %s <fir|conv2d> <1t|mt> <iters> [n|side]\n", argv[0]);
        return 2;
    }
    const int iters = std::atoi(argv[3]);
    if (iters < 1) {
        std::fprintf(stderr, "iters must be >= 1\n");
        return 2;
    }
    const bool multithreaded = std::strcmp(argv[2], "mt") == 0;
    const bool fir = std::strcmp(argv[1], "fir") == 0;
    const size_t default_n = fir ? 65536 : 512;
    const size_t n = argc == 5 ? static_cast<size_t>(std::atoll(argv[4])) : default_n;
    if (fir)
        run_fir(multithreaded, iters, n);
    else
        run_conv(multithreaded, iters, n);
}

File: benches/shapes/shapes_baseline.rs
// CPU baselines for the fir/conv2d Mapal shapes, size-parameterized (S29 scale-up).
// Build: rustc -O -C target-cpu=native shapes_baseline.rs -o shapes_baseline_rs
use std::time::Instant;

fn cgroup_quota() -> Option<usize> {
    std::fs::read_to_string("/sys/fs/cgroup/cpu.max")
        .ok()
        .and_then(|text| {
            let mut fields = text.split_whitespace();
            let quota = fields.next()?;
            let period = fields.next()?.parse::<usize>().ok()?;
            if quota == "max" || period == 0 {
                return None;
            }
            quota
                .parse::<usize>()
                .ok()
                .map(|quota| quota.div_ceil(period).max(1))
        })
        .or_else(|| {
            let quota = std::fs::read_to_string("/sys/fs/cgroup/cpu/cpu.cfs_quota_us")
                .ok()?
                .trim()
                .parse::<isize>()
                .ok()?;
            let period = std::fs::read_to_string("/sys/fs/cgroup/cpu/cpu.cfs_period_us")
                .ok()?
                .trim()
                .parse::<usize>()
                .ok()?;
            (quota > 0 && period > 0).then(|| (quota as usize).div_ceil(period).max(1))
        })
}

fn thread_width() -> usize {
    if let Some(threads) = std::env::var("THREADS")
        .ok()
        .and_then(|value| value.parse::<usize>().ok())
        .filter(|&threads| threads >= 1)
    {
        return threads;
    }
    let available = std::thread::available_parallelism().map(usize::from).unwrap_or(1);
    cgroup_quota().map_or(available, |quota| available.min(quota))
}

fn fir_range(x: &[f32], w: &[f32], y: &mut [f32], begin: usize) {
    for (offset, value) in y.iter_mut().enumerate() {
        let t = begin + offset;
        let mut acc = 0.0f32;
        for k in 0..w.len() {
            acc += w[k] * x[t + k];
        }
        *value = acc;
    }
}

fn conv_rows(img: &[f32], w: &[f32], out: &mut [f32], side: usize, row_begin: usize) {
    let stride = side + 2;
    for (row_offset, row) in out.chunks_exact_mut(side).enumerate() {
        let i = row_begin + row_offset;
        for (j, value) in row.iter_mut().enumerate() {
            let mut acc = 0.0f32;
            for k in 0..9 {
                acc += w[k] * img[(i + k / 3) * stride + j + k % 3];
            }
            *value = acc;
        }
    }
}

fn run_fir(multithreaded: bool, iters: usize, n: usize) {
    let x: Vec<f32> = (0..n + 63)
        .map(|t| (((t * 7 + 13) % 101) as i32 - 50) as f32)
        .collect();
    let w: Vec<f32> = (0..64)
        .map(|k| (((k * 5 + 3) % 31) as i32 - 15) as f32)
        .collect();
    let mut y = vec![0.0f32; n];
    let threads = thread_width().min(y.len());
    let chunk = y.len().div_ceil(threads);

    for _ in 0..iters {
        let start = Instant::now();
        if multithreaded {
            std::thread::scope(|scope| {
                for (lane, slice) in y.chunks_mut(chunk).enumerate() {
                    let x = &x;
                    let w = &w;
                    scope.spawn(move || fir_range(x, w, slice, lane * chunk));
                }
            });
        } else {
            fir_range(&x, &w, &mut y, 0);
        }
        println!("iter ms={:.6}", start.elapsed().as_secs_f64() * 1000.0);
    }
    println!("{}\n{}", y[0], y[n - 1]);
}

fn run_conv(multithreaded: bool, iters: usize, side: usize) {
    let stride = side + 2;
    let img: Vec<f32> = (0..stride * stride)
        .map(|t| (((t * 7 + 13) % 101) as i32 - 50) as f32)
        .collect();
    let w: Vec<f32> = (0..9)
        .map(|k| (((k * 5 + 3) % 31) as i32 - 15) as f32)
        .collect();
    let mut out = vec![0.0f32; side * side];
    let threads = thread_width().min(side);
    let rows_per_thread = side.div_ceil(threads);
    let chunk = rows_per_thread * side;

    for _ in 0..iters {
        let start = Instant::now();
        if multithreaded {
            std::thread::scope(|scope| {
                for (lane, slice) in out.chunks_mut(chunk).enumerate() {
                    let img = &img;
                    let w = &w;
                    scope.spawn(move || conv_rows(img, w, slice, side, lane * rows_per_thread));
                }
            });
        } else {
            conv_rows(&img, &w, &mut out, side, 0);
        }
        println!("iter ms={:.6}", start.elapsed().as_secs_f64() * 1000.0);
    }
    println!("{}\n{}", out[0], out[side * side - 1]);
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if !(4..=5).contains(&args.len())
        || !matches!(args[1].as_str(), "fir" | "conv2d")
        || !matches!(args[2].as_str(), "1t" | "mt")
    {
        eprintln!("usage: {} <fir|conv2d> <1t|mt> <iters> [n|side]", args[0]);
        std::process::exit(2);
    }
    let iters = args[3]
        .parse::<usize>()
        .ok()
        .filter(|&value| value >= 1)
        .unwrap_or_else(|| {
            eprintln!("iters must be >= 1");
            std::process::exit(2);
        });
    let default_n = if args[1] == "fir" { 65536 } else { 512 };
    let n = args
        .get(4)
        .and_then(|value| value.parse::<usize>().ok())
        .unwrap_or(default_n);
    if args[1] == "fir" {
        run_fir(args[2] == "mt", iters, n);
    } else {
        run_conv(args[2] == "mt", iters, n);
    }
}
